### ISSUE TITLE
Allow paired-end mapping for STARsolo

### ISSUE THREAD

**Issue by rspreafico-vir** [alexdobin/STAR#768]
Hi there, thanks for the awesome STARsolo tool and the recently released improvements!

The read structure for 10X Genomics 5' gene expression libraries is such that both Read 1 and Read 2 carry cDNA information:

https://assets.ctfassets.net/an68im79xiti/2mhX7BBB4s4iGEYUB6jzN6/5c4c24ecaa2c3af1cdf95e8e14424504/CG000207_ChromiumNextGEMSingleCellV_D_J_ReagentKits_v1.1_UG_RevD.pdf

Note the link is primarily for VDJ libraries, but it applies to GEX too (the insert will just be any cDNA rather than TCR/BCR). One could split Read 1 in two segments: one containing CB, UMI and TSO, and the other containing cDNA (5' of fragment). Conversely, read 2 is purely cDNA (3' of fragment). 

Currently STARsolo allows to map only single-end cDNA reads - the other read is assumed to be used wholly for barcodes. Would it be possible to allow mapping of paired end cDNA reads?

Thank you

**Comment by rspreafico-vir** [comment]
Another link to read structure is the following:

https://kb.10xgenomics.com/hc/en-us/articles/360000939852-What-is-the-difference-between-Single-Cell-3-and-5-Gene-Expression-libraries-

Note that often sequencing providers sequence at 2x150, that is the reason for extra sequence information on read 1 other than barcodes. 

By contrast, using read 1 in 3' GEX libraries is more complicate due to presence of varying-length poly-T. In 5' GEX libraries, this is not a concern: after barcodes, a fixed-length TSO is present, rather than a variable-length polyT, so it can be more easily trimmed out. 

Having read 1 mapped along with read 2 would increase mapping accuracy and would also provide exact 5' start site of transcript.

Thanks

**Comment by alexdobin** [comment]
Hi Roberto,

this should be doable, thanks for the suggestion! Is there a dataset that was sequenced that way, with cDNA in both read1 and 2? I will be able to work on it next month.

Cheers
Alex

**Comment by rspreafico-vir** [comment]
I do have one but unlikely that I will get permission to share. In principle, this type of dataset should become (or have become) more common as people may not strictly adhere to 10x custom cycling directions (e.g. 26+98 cycles) and rather go for something more standard (e.g. 150+150, e.g. to multiplex on the Novaseq). That creates the opportunity to have cDNA sequence information of both mates. This is particularly easy in 5' GEX libraries, but it is also possible in 3' GEX libraries once the polyA/T is removed.

I found this link illustrating read structure particularly well done:

https://teichlab.github.io/scg_lib_structs/

Inquired with 10x Genomics about it as well, and turns out that Cell Ranger has a "PE" mode that gets activated when the barcode read (read 1) is longer than 50 nt - and so it gets used for both cDNA and barcode.

This means that either of the following would need to happen:
- STAR gets fed 3 reads simultaneously: Read1-cDNA, Read2-cDNA and Read 1-Barcode (the user would take care of splitting read 1 in two chunks: barcode and cDNA, and in the process remove garbage such as TSO sequence that does not belong to either)
- STAR gets fed 2 reads and the user has a way to flexibly specify CB, UMI, cDNA forward and cDNA reverse coordinates

Thanks Alex for looking into this, very much appreciated!

**Comment by alexdobin** [comment]
Hi Roberto,

I would need to have a dataset to test it on. Could you, simulate a ~1M reads from a dataset that you have in the following way: replace the cDNA portions of the reads that you have with some human sequences (say from some public PE dataset), while keeping the CB and UMI intact. This will give me enough reads to work with.

Cheers
Alex

**Comment by rspreafico-vir** [comment]
Cool! Will do one of these days. Do you have a box I can drop these files into?

**Comment by jenzopr** [comment]
Hi Alex,

PE-Support would be great! Also for the scATAC-seq protocol from 10X (e.g. output from [cellranger-atac mkfastq](https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/using/mkfastq)). You can find example files also on the link posted (i.e. [tiny-bcl tar file](http://cf.10xgenomics.com/supp/cell-atac/cellranger-atac-tiny-bcl-1.0.0.tar.gz))

Best and looking forward to test star-solo!

**Comment by alexdobin** [comment]
Hi Jens,

STARsolo can map and error-correct cell barcodes for PE reads (e.g. scATAC-seq), please see this thread for details: https://github.com/alexdobin/STAR/issues/558

Cheers
Alex

**Comment by jenzopr** [comment]
Alex, thats great! Thanks 👍

**Comment by Bernadetadad** [comment]
Hi Alex, any update on this issue? If you still need a dataset I have one that would probably work for this and I'd be happy to share it.

**Comment by alexdobin** [comment]
Hi Bernadeta,

how are your reads formatted? There is no standard format for PE single cell reads, STARsolo may be able to support some formats - probably with some preprocessing of the reads.

Cheers
Alex

**Comment by DarioS** [comment]
There is now [a publicly available one](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926) of COVID-19 patients. Code to download FASTQ:

```
accessionsToGet=(SRR11181954 SRR11181955 SRR11181956 SRR11181957 SRR11181958 SRR11181959\
                 SRR11537946 SRR11537947 SRR11537948 SRR11537949 SRR11537950 SRR11537951)

for accession in ${accessionsToGet[@]}
do
  fasterq-dump $accession -O /home/datasets/COVID_single-cell/reads/ --split-spot --split-files
done
```

**Comment by alexdobin** [comment]
Hi Dario,

can you describe the formatting of the reads in this protocol - i.e. locations/lengths of CB, UMI and cDNA sequences?

Cheers
Alex

**Comment by DarioS** [comment]
I don't know, but I read "Chromium Single Cell 5’ library preparation kit according to the manufacturer’s protocol (10x Genomics)." I'm a beginner so I couldn't make a good educated guess about it.

**Comment by benplm** [comment]
Hi, I jump into the thread since I think I have the same library preparation as DarioS (i.e. Chromium Single Cell 5' Next Gem v1.1 for immune profiling + gene expression).

The layout is paired-end 150bp mates where read1 contains cell barcode (16bp) + UMI (10 bp) + TSO (13 bp) + first mate of the cDNA (111 bp), while read2 contains only the second mate of the cDNA (150 bp). Right now paired-end alignment is only possible with cell ranger, as far as I can tell.

10x distributes some example datasets: https://support.10xgenomics.com/single-cell-vdj/datasets

Hope this can help.

**Comment by benstory** [comment]
Any update on this? Here is another dataset with the same construction: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE159929

**Comment by alexdobin** [comment]
Hi @benstory 

no update - but thanks for bumping it. I will try to include this capability in the next release, ETA ~ this week.

Cheers
Alex

**Comment by alexdobin** [comment]
Hi @benstory 

this feature was implemented in https://github.com/alexdobin/STAR/releases/tag/2.7.8a,
details: https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#barcode-and-cdna-on-the-same-mate


Cheers
Alex

**Comment by benstory** [comment]
Awesome thank you!

**Comment by Terkild** [comment]
I cannot get this to work by following the "details" link above. I get the error:

"EXITING because of fatal PARAMETER error: --clip5pAdapterSeq has to contain 2 values to match the number of mates.
SOLUTION: specify 2values in --clip5pAdapterSeq , for no clipping use -"

If i try to set "--clip5pAdapterSeq 39 0" then I get an error that it has not been implemented yet.

This is my STAR call:

STAR --runMode alignReads 
--runThreadN 16 
--genomeDir /data/ref/STAR/cellranger-GRCh38-2020-A 
--soloCBwhitelist /opt/cellranger-6.0.0/lib/python/cellranger/barcodes/737K-august-2016.txt 
--soloType CB_UMI_Simple     
--outFileNamePrefix STAR_out/ 
--outSAMtype BAM SortedByCoordinate 
--outSAMattributes NH HI nM AS CR UR CB UB GX GN sS sQ sM     
--readFilesCommand zcat 
--clipAdapterType CellRanger4 
--outFilterScoreMin 30 
--soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts 
--soloUMIfiltering MultiGeneUMI_CR 
--soloUMIdedup 1MM_CR     
--soloBarcodeMate 1   
--clip5pNbases 39 0 
--soloCBstart 1 
--soloCBlen 16 
--soloUMIstart 17 
--soloUMIlen 10     
--readFilesIn read1.fastq.gz read2.fastq.gz

Are some of my additional parameters incompatible with using the chip5pNbases command?

**Comment by alexdobin** [comment]
Hi Terkild,

```--clip5pNbases``` cannot be used with ```--clipAdapterType CellRanger4```. The latter option is specific to 10X adapter sequence and should only be used for 10X data.

Cheers
Alex

**Comment by apredeus** [comment]
Hello Alex,

I wanted to point out that README probably should be updated for PE mode (I'm trying to implement a script that would pretty much mimic Cell Ranger behaviour for all possible cases, thus, I wanted to enable paired-end processing of 5' 10x data). 

So, when using the recommended settings, R1 and R2 are switched compared to what's normally used for 10x, and `--soloStrand Forward` needs to be used with 10x 5' data. If I use ` --soloStrand Reverse` (which works perfectly for the same sample in single-end mode), only few percent of reads are assigned to exons (as would be expected). 

Thus, the options that give highly similar results for PE and SE alignment/quantification of the same 5' 2x150 bp sample are:

```
STAR --runThreadN $CPUS --genomeDir $REF --readFilesIn $R1 $R2 --runDirPerm All_RWX --readFilesCommand zcat $SORTEDBAM \
     --soloBarcodeMate 1   --clip5pNbases 39 0 --soloStrand Forward \
     --soloType CB_UMI_Simple --soloCBwhitelist $BC --soloCBstart 1 --soloCBlen 16 --soloUMIstart 17 --soloUMIlen 10 \ 
     --soloUMIdedup 1MM_CR --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR \
     --soloCellFilter EmptyDrops_CR --outFilterScoreMin 30 \
     --soloFeatures Gene GeneFull Velocyto --soloOutFileNames output/ genes.tsv barcodes.tsv matrix.mtx
```
(for paired-end)

```
STAR --runThreadN $CPUS --genomeDir $REF --readFilesIn $R2 $R1 --runDirPerm All_RWX --readFilesCommand zcat $SORTEDBAM \
     --soloType CB_UMI_Simple --soloCBwhitelist $BC --soloBarcodeReadLength 0 --soloUMIlen 10 --soloStrand Reverse \
     --soloUMIdedup 1MM_CR --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR \
     --soloCellFilter EmptyDrops_CR --clipAdapterType CellRanger4 --outFilterScoreMin 30 \
     --soloFeatures Gene GeneFull Velocyto --soloOutFileNames output/ genes.tsv barcodes.tsv matrix.mtx
```    
(for single-end)

**Comment by bepoli** [comment]
I also had the same problem with 10x 5' data being reverse stranded, and solved once and for all (hopefully!) by checking this file in Cell Ranger's code: https://github.com/10XGenomics/cellranger/blob/master/lib/python/cellranger/chemistry.py

This file contains the settings specific for each 10x chemistry.
As you can see, SC5P_R2 chemistry's strandness is "-", while the strandness of SC5P_PE is "+". I guess this depends on the different library preparation strategy of the two kits.

So basically I ran 3' single-end data on forward, and 5' single-end data on reverse (both with R2 R1 files order), and got the expected number of counts in both (on 10x pbmc data).
However I didn't do some serious testing on this matter (i.e. I didn't ran the same datasets with both forward and reverse options and then checked which had the greater counts), but I'll let you know if I do.
